daffodil-runtime1/src/main/scala/edu/illinois/ncsa/daffodil/processors/parsers/PackedBinaryTraits.scala

```diff
@@ -0,0 +1,195 @@
+ *     this list of conditions and the following disclaimers.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimers in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the names of Tresys Technology, nor the names of its contributors
+ *     may be used to endorse or promote products derived from this Software
+ *     without specific prior written permission.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE
+ * SOFTWARE.
+ */
+
+package edu.illinois.ncsa.daffodil.processors.parsers
+
+import edu.illinois.ncsa.daffodil.processors.ElementRuntimeData
+import edu.illinois.ncsa.daffodil.processors.ParseOrUnparseState
+import edu.illinois.ncsa.daffodil.processors.FieldDFAParseEv
+import edu.illinois.ncsa.daffodil.processors.TextJustificationType
+import edu.illinois.ncsa.daffodil.processors.dfa.TextDelimitedParserBase
+import edu.illinois.ncsa.daffodil.util.Maybe
+import edu.illinois.ncsa.daffodil.exceptions.Assert
+import edu.illinois.ncsa.daffodil.equality._
+import edu.illinois.ncsa.daffodil.processors.dfa
+import edu.illinois.ncsa.daffodil.util.MaybeChar
+import edu.illinois.ncsa.daffodil.schema.annotation.props.gen.BinaryNumberCheckPolicy
+import java.math.{ BigInteger => JBigInteger, BigDecimal => JBigDecimal }
+import java.nio.charset.StandardCharsets
+import passera.unsigned.ULong
+
+trait PackedBinaryConversion {
+  def toBigInteger(num: Array[Byte]): JBigInteger
+  def toBigDecimal(num: Array[Byte], scale: Int): JBigDecimal
+}
+
+abstract class PackedBinaryDecimalBaseParser(
+  e: ElementRuntimeData,
+  binaryDecimalVirtualPoint: Int)
+  extends PrimParserObject(e)
+  with PackedBinaryConversion {
+
+  protected def getBitLength(s: ParseOrUnparseState): Int
+
+  def parse(start: PState): Unit = {
+    val nBits = getBitLength(start)
+    if (nBits == 0) return // zero length is used for outputValueCalc often.
+    val dis = start.dataInputStream
+
+    if (!dis.isDefinedForLength(nBits)) {
+      PE(start, "Insufficient bits in data. Needed %d bit(s) but found only %d available.", nBits, dis.remainingBits.get)
+      return
+    }
+
+    try {
+      val bigDec = toBigDecimal(dis.getByteArray(nBits, start), binaryDecimalVirtualPoint)
+      start.simpleElement.overwriteDataValue(bigDec)
+    } catch {
+      case n: NumberFormatException => PE(start, "NumberFormatException in packed data: \n%s", n.getMessage())
+    }
```

Comment:
This is really minor, but it's probably more correct to move the overwriteDataValue outside of the try/catch block. That function shouldn't throw an NFE, and if it did, it's probably a bug that we do not want to hide in a ParseError. So instead, you can do something like this:
```scala
val bigDec = try {
  toBigDecimal(...)
} catch {
  case ...
}
start.simpleElement.overwriteDataValue(bigDec)
```
Also, it's generally good to avoid mentioning interal class names like NumberFormatException where possible, just because that doesn't mean anything to a user. Can the error message be changed to something a little more user friendly? Maybe just "Error in packed data:"?

Reply:
I originally had that call outside the try block, but then bigDec was out of scope.  Hadn't thought of using the ```val bigDec = try {...}``` type of declaration, which still feels strange to me coming from C/C++ and their very simple declarations.

Reply:
Yeah, it definitely feels weird at first, but is pretty standard syntax in Scala. It's also useful for things like if-statements, e.g.:
```scala
val foo = if (condition) {
  "bar"
} else {
  "baz"
}
```

Reply:
After trying this out and talking to Steve, because ```PE(...)``` no longer throws, I would need to provide some sort of sane default for bigDec which isn't really possible.  So we decided to just leave the call to ```start.simpleElement.overwriteDataValue(bigDec)``` inside the try block.